RX - Improved RX failsafe detection & handling

ProDrone opened the pull request:
Supersedes #1361
 
To solve problems as indicated here:
https://github.com/cleanflight/cleanflight/issues/1266#issuecomment-135640133

and here:
https://github.com/cleanflight/cleanflight/pull/1340

and here:
https://github.com/cleanflight/cleanflight/pull/1342

Tested on FrSKY X4RSB with latest CPPM firmware (non-EU version).
Firmware filename: X4R-X4RSB_cppm_non-EU_150630

In both SBUS and CPPM mode.

---
Added CLI variable `rxfail_delay` to rxfail detection

Used when out of range pulses from the receiver are detected.
Delay in deciseconds before rxfail values are applied (channel value is on HOLD until then).
Default 3 (range: 3 - 10).

This should prevent a too aggressive reaction to small dropouts.

hydra commented:
Great @ProDrone. Let's get some feedback and we'll get this merged to master.  I think it's probably worth creating a new branch 1.10.1 based on the 1.10.0 tag and applying this to it and releasing new binaries (with any other fixes that come up in the next few days)

borisbstyle commented:
@ProDrone 
Added this one and will test soon

ProDrone commented:
@borisbstyle Thanks. Looking forward to some feedback.

hydra commented:
@borisbstyle Did you manage to do some testing of it?

borisbstyle commented:
Well yes and no. I have it loaded on my code, but I kind of forgot what
would it do.
Many betaflight people still suffer from glitches like spektrum rx. The
same issue is there on 1.10 as well. I let them verify.
I initially thought that this pr would fix that.
Op 12 okt. 2015 21:04 schreef "Dominic Clifton" <notifications@github.com>:

> @borisbstyle <https://github.com/borisbstyle> Did you manage to do some
> testing of it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cleanflight/cleanflight/pull/1374#issuecomment-147492691>
> .
>

ProDrone commented:
@borisbstyle The code in betaflight is a modified version:
Changes are in rx.c
https://github.com/borisbstyle/cleanflight/blob/betaflight/src/main/rx/rx.c#L392-L395
Which will allow aux channels go to midrc on glitches. Maybe cause disarm when arming with switch?
Changes in failsafe.c
https://github.com/borisbstyle/cleanflight/blob/betaflight/src/main/flight/failsafe.c#L168-L169
In effect bypasses the 200ms minimum delay for RX link_up & link_down, and will start failsafe handler immediately.

Can you please use the original code in betaflight?

This PR aims at making the detection of failsafe event better.

I will add more code shortly (intended tomorrow) in an attempt to prevent glitches during normal operation. The problem is that i do not see these glitches during my testing here... (yet).

borisbstyle commented:
@ProDrone 
This is where I was talking about. This guy tested on 1.10 after my request, but he should have been testing 1.11 master with this PR?
http://www.rcgroups.com/forums/showpost.php?p=32918829&postcount=5572

The problem with failsafe code was that it just simply was causing issues a while ago and was modified therefore. I was not able to find a problem of not working failsafe in some scenarios and bypassed that part, which worked in all scenarlos. Rather failsafe than sorry. Maybe it should be working now again after all other latest fixes.

The thing you mention about midrc is to give users same behaviour like on previous 1.9. In one of the previous PR's it wasn't the case and AUX channels were flickering to LOW state, which often was disarm/arm. Maybe that can be changed back now.

Also a lot of "faulty" X4RSB behaviours were also not helping me in good failsafe testing. 
You do have to understand that failsafe part is perhaps the most discussed and under pressure in the betaflight thread. It is really a pain for me to start modifying those again now.

borisbstyle commented:
Ok added back original failsafe code as in current master.

I also let the guy test again and he only seems to have problem with spektrum 1024, but not on 2048. I asked him to open an issue. He has the same issue on 1.10 so it has nothing to do with this PR, but you can have a look anyway.

ProDrone commented:
> This is where I was talking about. This guy tested on 1.10 after my request, but he should have perhaps been testing 1.11 master with this PR?
http://www.rcgroups.com/forums/showpost.php?p=32918829&postcount=5572

1.11 does not yet has this PR. So that would give the same result.

> The thing you mention about midrc is to give users same behaviour like on previous 1.9. In one of the previous PR's it wasn't the case and AUX channels were flickering to LOW state, which often was disarm/arm. Maybe that can be changed back now.

I still remember when you did put this in and for what problem that was :grinning: These problems were during startup due to wrong initialization (solved in #1351).

> Also a lot of "faulty" X4RSB behaviours were also not helping me in good failsafe testing.
You do have to understand that failsafe part is perhaps the most discussed and under pressure in the betaflight thread. It is really a pain for me to start modifying those again now.

I do understand your position, but to improve it has to change anyway.

I prefer to use the same code for betaflight and cleanflight. rx fail and failsafe are highly configurable and for race quads i would recommend to set 
~~~
`rx fail` for stick channels 0 - 3 to hold,
`failsafe_delay` to 50,
`failsafe_off_delay` to 0,
`failsafe_throttle` to 1000.
~~~
That way you should not notice single signal drops (cause that is what it looks like) and after 700 ms (200ms fixed + 500ms delay) failsafe handling will activate and drop the quad followed by immediate disarm. That is what most fast-lane users prefer i think.

Dropouts can always occur but i feel the current `rx fail` system is to sensitive and i will fix that soon.

ProDrone commented:
@borisbstyle did you add this PR to it?

ProDrone commented:
@borisbstyle Oops i meant `failsafe_delay to 5` above...

borisbstyle commented:
yes it was already added :)

But now 1:1 equal failsafe handling as well.
Op 13 okt. 2015 01:02 schreef "ProDrone" <notifications@github.com>:

> @borisbstyle <https://github.com/borisbstyle> Oops i meant failsafe_delay
> to 5 above...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cleanflight/cleanflight/pull/1374#issuecomment-147544493>
> .
>

borisbstyle commented:
@ProDrone Can you perhaps prioritize rxfail erratic behaviour. It think that's the real problem we have anyway. It should really work with steps to it's goal instead of abrupt channel changes.

borisbstyle commented:
@ProDrone 
Your suggestion for failsafe parameters is only for 0-3 channels. What about users arming/disarming on aux channel? To set those to "h" as well?

Funny enough all the reports I had with quads dropping out of sky are all users who use switch arming. I would say the glitch made their quads to disarm probably.

borisbstyle commented:
https://github.com/cleanflight/cleanflight/issues/1394

borisbstyle commented:
@ProDrone 
What's the reason you are using different datatype for this function? rxdata is int16_t and you went for uint16_t?
https://github.com/cleanflight/cleanflight/blob/master/src/main/rx/rx.c#L375

ProDrone commented:
@borisbstyle 

> Can you perhaps prioritize rxfail erratic behaviour. It think that's the real problem we have anyway. It should really work with steps to it's goal instead of abrupt channel changes.

Don't worry, it is on top of my list!

Making it work more smooth is another thing which is already on my list for some time now, see: https://github.com/cleanflight/cleanflight/pull/995#issuecomment-127578832 It is just that the detection must work good before implementing that part, otherwise it may hide other problems.

At the moment it is still unclear why these short drops occur, this needs more investigation.

> Your suggestion for failsafe parameters is only for 0-3 channels. What about users arming/disarming on aux channel? To set those to "h" as well?

__Yes__, an aux channel used for arming can best be set to hold `rxfail h`.
Single signal drops can be covered with that. It is important to combine this with an active failsafe, which should do the handling of persistent failure (mainly disarm). Otherwise it will be a potential fly away...

Set `rxfail s` is best to be used for channels used for a lost-and-found HK type beeper or lights.

ProDrone commented:
> What's the reason you are using different datatype for this function? rxdata is int16_t and you went for uint16_t?
https://github.com/cleanflight/cleanflight/blob/master/src/main/rx/rx.c#L375

No reason in particular, i probably adapted to `uint16_t sample` and the returned type from function
`uint16_t  calculateNonDataDrivenChannel()`, it was there for some time already, see:
https://github.com/cleanflight/cleanflight/blob/v1.7.2/src/main/rx/rx.c#L301-L316
It is not as clean as i it can be, but it doesn't hurt either.

borisbstyle commented:
@ProDrone I think I have some test code for smooth rx_fail transition. I think it might be code after some cleanup. Will show you soon as I publish it somewhere.
I think every scenario is covered. You need to review and see what of that we could use for final version.

borisbstyle commented:
@ProDrone 
Check this PR in my own branch I made to discuss. Quick and dirty, but tested. You get the picture
https://github.com/borisbstyle/cleanflight/pull/19

ProDrone commented:
@borisbstyle Before applying some sort of smoothing i first wanted suppression of too fast rxfail triggering. It should not be hidden by another mechanism in my view. I am a bit surprised that you started working on the same subject :yum:. I strongly feel that we should field test these things step-by-step and not put it all in at once. I will look at your code now...

borisbstyle commented:
I am not working on it at all. This is just an experiment and not included
anywhere yet. I just made the code available on this way.
It didnt cost me a lot of time but I learned my way through failsafe code.
Op 14 okt. 2015 19:49 schreef "ProDrone" <notifications@github.com>:

> @borisbstyle <https://github.com/borisbstyle> Before applying some sort
> of smoothing i first wanted suppression of too fast rxfail triggering. It
> should not be hidden by another mechanism in my view. I am a bit surprised
> that you started working on the same subject [image: :yum:]. I strongly
> feel that we should field test these things step-by-step and not put it all
> in at once. I will look at your code now...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cleanflight/cleanflight/pull/1374#issuecomment-148133404>
> .
>

ProDrone commented:
Okay that is a good thing, learning the code always is. I am just a bit worried that we end up with several different approaches to solve the same problem which may cause confusion.

Do you agree then that step-by-step testing is the best solution?
I expect (and hope) the latest fix i just pushed should work fine.
_(i hate those movies with happy flying quads that end with a free fall...)_

borisbstyle commented:
Yes I just wanted to understand it better.

I do agree with @ledvinap that less complicity is better. Another added
special scenario would only add more overhead.

Step by step is indeed a good approach, but there should be a good idea
behind it all like end-to-end design to work towards that.
I see many separate fixes on fixes lately and I am afraid that we loose the
entire picture on this way and get too much checks on checks.

Also the failsafe configuration should be easier or better documented
towards endusers.

But to the point now.
Why do we need rxfail at all on the eay it is now? Maybe rxfail should be
the failsafe configuration for all channels instead of intermediate
failsafe. We are able to detect all types of failsafes right?
Op 14 okt. 2015 20:19 schreef "ProDrone" <notifications@github.com>:

> Okay that is a good thing, learning the code always is. I am just a bit
> worried that we end up with several different approaches to solve the same
> problem which may cause confusion.
>
> Do you agree then that step-by-step testing is the best solution?
> I expect (and hope) the latest fix i just pushed should work fine.
> *(i hate those movies with happy flying quads that end with a free
> fall...)*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cleanflight/cleanflight/pull/1374#issuecomment-148143555>
> .
>

borisbstyle commented:
Ok I agree with this approach. I will already bring it in betaflight. Let's see if 3 fails will be enough.

borisbstyle commented:
So what do you think @prodrone?  I personally think we might need to apply timer here rather than counts to not cause huge timing differences in different receivers.
Apparantly we dont live in a perfect world and one receiver drops more than other.

Maybe this should be part of failsafe_delay?

ProDrone commented:
@borisbstyle 
> I do agree with @ledvinap that less complicity is better. Another added special scenario would only add more overhead.

I like simple too, simple is good, too simple is bad. There are a lot of cases to cover, so many different receivers and configurations. At the level of failsafe handling we like everything to behave the same but at the driver level all is completely different. That is why there are so many exceptions. 

For example: An FrSKY X4R-SB set to sent no pulses, keeps sending serial frames with 880us pulses, a Graupner GR-16 simply stops serial transmission for the same.

This code can be taken apart or completely rewritten, but that will undoubtedly introduce new faults. I like _if it works don't fix it_ here.

> ...but there should be a good idea behind it all like end-to-end design to work towards that. I see many separate fixes on fixes lately and I am afraid that we loose the entire picture on this way and get too much checks on checks.

Don't worry, there is a good idea behind it, but just no formal design specification.

> Also the failsafe configuration should be easier or better documented towards endusers.

Completely agree, i am just not that good at producing documentation, maybe we should ask @joshuabardwell for help with that? He seems to always find the right words when explaining something.

> Why do we need rxfail at all on the eay it is now? Maybe rxfail should be the failsafe configuration for all channels instead of intermediate failsafe. We are able to detect all types of failsafes right?

__rxfail__ is __stage_1__ of failsafe. It should cover the period when there is still a change at full recovery from a possible failsafe event. It corrects single channel errors and prepares all channels for stage_2.

__Failsafe__ handling is __stage_2__. This is for now just simple landing or dropping, but may become a full RTH function in the near future.

> So what do you think @prodrone? I personally think we might need to apply timer here rather than counts to not cause huge timing differences in different receivers.
Apparantly we dont live in a perfect world and one receiver drops more than other.

> Maybe this should be part of failsafe_delay?

An error counter is not a bad thing by itself. Maybe this should only apply to serial RX's.

Not too long ago i adapted the whole failsafe system which was based on counters. It is now based on time. So, yes, time is important here. I was thinking of subtracting the time, spend on bad puls detection, from the `failsafe_delay`.

I have (obviously) more thought on this subject, but like to see if this PR has stopped glitches and the dropping quad issues first.

Did you only receive positive feedback from testing this PR so far?

borisbstyle commented:
@ProDrone Yes everyone I spoke via PM's and in thread who was experiencing problems gave positive feedback. No doubt about it that it's covered now.

We went for 3, than for 6 and it was still there on spektrums which seem to be the worst case scenario from the feedback we have. So I decided if its there on 6 it will probably be too tight to chose 10 as a value so I went for safety margin and did 20. 20 should be around 200ms for serial.

I would like to see 200ms timer there, which will work great for all protocols and be consistent on them all.

I do think this needs to be solved fast and a 1.10.1 or 1.11 should be released.

hydra commented:
@borisbstyle @ProDrone ok, so lots of talk in here.  Do any more commits need to be made to this PR before I review it?

ProDrone commented:
@hydra Please review it and since it already is a good counter measure, it would be great to release it too. I will make additional modifications later on, unless you want to wait for that, i will do so tomorrow and the weekend.

borisbstyle commented:
yes this should solve it all from the feedback I had.

hydra commented:
i will wait for ProDone to do cleanup work.

ProDrone commented:
@hydra Adapted, tested, squashed, rebased on latest (tested again), first post updated, see commit message for more details.

- It is now time based instead of counter based.
- Collision with rc_smoothing function resolved.

- Initialize ARM switch to OFF position when arming via switch is defined.
To prevent arming during init when RX is disconnected and/or when RX is
connected but TX is still off.

- Documentation updated.
- Unit test adapted.

- CLI variable `rxfail_delay` added. I think it is required for a while, to prevent updates when time-out requirements change (receiver dependent).

borisbstyle commented:
IMO failsafe_delay should be enough? Any reason for not doing it?

This is how it is now:
Normal Operation->rxfail_delay->failsafe_delay->failsafe_off_delay->disarm

I think this is how it should be I guess:
Normal operation->failsafe_delay->failsafe_off_delay->disarm


The main reason I am saying this is because the enduser can't simply know what value to use for rxfail_delay without debugging and measuring and that requires coding by them self. It also opens a room for entering faulty values.
What do you think?

ProDrone commented:
This is how it works:

1. __state__  - normal operation
2. event - bad pulses received / rc link problems
3. __state__  - monitor and tolerate these for some time, keeping channels on position hold.
4. event - guard time `rxfail_delay` expired.
5. __state__  - replace channel values according to `rxfail` settings, notify failsafe handler of problems.
6. event - guard time `failsafe_delay` expired (problems persist, rc link has not recovered).
7. __state__  - start executing the failsafe procedure (anything like, drop death, auto-land, RTH).

That is why there are two timers. What i like to do is hide most of this from end users and update the documentation so it provides a good explanation, the lack of that causes these questions i think. But since there are still some more changes to do i postpone that a little.

This PR started with a counter and is now time based. This is more deterministic but it can be that it does not behave the same as the (now successful) counter based version.

hydra commented:
@ProDrone Sorry I missed you on IRC earlier tonight.  Thanks for giving me a heads-up that you've worked on it.  I will attend to this issue asap.

borisbstyle commented:
@ProDrone 
Would just hardcoding the value to 200ms be enough? That would prevent the confusion unless you know a scenario where it needs to be set to higher timing?
Failsafe_delay also has hardcoded 200ms I see. That could get removed now I believe?

ProDrone commented:
@borisbstyle I think you didn't read the commit message (at the end) :smile:

It is the intention to remove this asap. I had it at 300ms fixed at first, but realized that there is no guarantee that a counter of 20 (like it now in the field) couldn't take longer. Depends on your receiver. __The effect would be glitches again.__ :scream:

If and when that happens we need the ability to instruct users to set the timing higher. Until then this setting can be left at default.

borisbstyle commented:
@ProDrone Btw I missed this. Before rcData was never altered in the rx code. Now we are doing this on more places and perhaps it's better to move rcSmoothing to the rccommand to keep it more clean/readible?

borisbstyle commented:
I do think that 300ms will cover all cases. Remember it was only the serials showing this problem. Those will stay within that range. And even 20 is a safe margin above the worst case scenarios, which was spektrum sattellite. What do you think?

ProDrone commented:
> Btw I missed this. Before rcData was never altered in the rx code. Now we are doing this on more places and perhaps it's better to move rcSmoothing to the rccommand to keep it more clean/readible?

It is covered now, no need to change rcSmoothing.

borisbstyle commented:
@ProDrone 
It should be failry easy. Needs some testing. I think it is better perhaps?
https://github.com/cleanflight/cleanflight/pull/1418

ProDrone commented:
@borisbstyle Yes it looks better. I can remove the backup rcData after you confirmed that your modifications works fine. Are you going to test it now?

borisbstyle commented:
@ProDrone 
can't test right now (not at home). Wil do it this evening if you dont mind?

I personally think it is better approach to leave rcData untouched in rcSmoothing as it is static and now can conflict with new failsafe. I should have done this before.

borisbstyle commented:
@ProDrone You are also welcome to test it :dancer: 

It is basically the same logic but applied on different variable. Just needed to be moved to be applied after annexCode()

hydra commented:
@borisbstyle @ProDrone I will wait for more testing on this, sounds like you're almost there.  Please let me know when it's ready.

ProDrone commented:
@hydra Well, it works fine, but the question is if you agree with the added CLI variable `rxfail_delay` or find it better to put a fixed 300ms instead (it is just a precaution). I made a backup/restore for rcData since rcSmoothing interfered with that. Should i leave it in, wait for @borisbstyle confirmation on his fix or _assume_ it will work and remove the backup/restore upfront?

borisbstyle commented:
@ProDrone Just wait till I test it. Wont take that long. I dont like the idea of backup and restore just to not interfere other function later on the code. rcData should be as transparant as possible.

ProDrone commented:
Completely agree with removing backup/restore. I did that to make an independent change cause i did not want to modify your stuff (without being able to verify the outcome). I will wait.... :hourglass:

borisbstyle commented:
@ProDrone Can you already merge my PR into this one? You can already prepare your code in it.

ProDrone commented:
@borisbstyle I prefer to wait for your confirmation that it works, before making changes to this PR.

borisbstyle commented:
@ProDrone Ok

borisbstyle commented:
@ProDrone Ok tested. It works :D. Continue finishing this PR.

ProDrone commented:
@borisbstyle Great! I will remove the backup/restore stuff from this PR. Assuming that you make a separate PR fro your mods. I am tempted to also remove the CLI `rxfail_delay` (because less is better) and make it fixed to 300ms again, but like to know what @hydra thinks...

borisbstyle commented:
Just do what you think its best. My pr is finished. Same pr like I showed
you this afternoon. I only tested it.
Op 20 okt. 2015 20:41 schreef "ProDrone" <notifications@github.com>:

> @borisbstyle <https://github.com/borisbstyle> Great! I will remove the
> backup/restore stuff from this PR. Assuming that you make a separate PR fro
> your mods. I am tempted to also remove the CLI rxfail_delay (because less
> is better) and make it fixed to 300ms again, but like to know what @hydra
> <https://github.com/hydra> thinks...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cleanflight/cleanflight/pull/1374#issuecomment-149660637>
> .
>

ProDrone commented:
@hydra Finished, ready for merging, see commit message.

borisbstyle commented:
@ProDrone 
Looks good to me. I merged it into Betaflight as well, but it all looks fine to me and I agree with the approach.

borisbstyle commented:
@ProDrone 
There is just one scenario what is not yet fully covered. When rxfail kicks in during the flight it still can trigger the disarm even though the possibility that it will happen is now reduced by much.
I mean when users want to use full failsafe feature they would need to take account configuring their arming switch properly to not interfere with rxfail.
Other option is to do kind of same like you did with the init part, but opposite way around to keep it armed so the failsafe feature does it's job.

ProDrone commented:
@borisbstyle 
The channel used for the arming switch should be set to HOLD `rxfail h`. Then when rxfail is applied it will keep its armed/disarmed position. That value is also left alone when the failsafe handler takes over.

I think that will cover what you meant.

borisbstyle commented:
@ProDrone Yes you are right.

But no issues reported with this one and many people used it over the last week :D.

hydra commented:
Merging.
